lib/elixir/lib/calendar/datetime.ex

```diff
@@ -172,20 +172,213 @@ defmodule DateTime do
+
+    case validate_positive_leap_second(dt, tz_db_or_config) do
+      :ok ->
+        {:ok, dt}
+
+      {:error, :outside_leap_second_data_range} ->
+        {:outside_leap_second_data_range, dt}
+
+      error ->
+        error
+    end
+  end
+
+  def from_naive(naive_datetime, "Etc/UTC", _) do
+    do_from_naive(naive_datetime, "Etc/UTC", 0, 0, "UTC")
+  end
+
+  def from_naive(%{calendar: Calendar.ISO} = naive_datetime, time_zone, tz_db_or_config) do
+    case TimeZoneDatabaseClient.time_zone_periods_from_wall_datetime(
+           naive_datetime,
+           time_zone,
+           tz_db_or_config
+         ) do
+      {:single, period} ->
+        do_from_naive_check_leap_second(
+          naive_datetime,
+          time_zone,
+          period.std_offset,
+          period.utc_offset,
+          period.zone_abbr,
+          tz_db_or_config
+        )
+
+      {:ambiguous, first_period, second_period} ->
+        {:ok, first_datetime} =
+          do_from_naive(
+            naive_datetime,
+            time_zone,
+            first_period.std_offset,
+            first_period.utc_offset,
+            first_period.zone_abbr
+          )
+
+        {:ok, second_datetime} =
+          do_from_naive(
+            naive_datetime,
+            time_zone,
+            second_period.std_offset,
+            second_period.utc_offset,
+            second_period.zone_abbr
+          )
+
+        {:ambiguous, first_datetime, second_datetime}
+
+      {:gap, {first_period, first_period_until_wall}, {second_period, second_period_from_wall}} ->
+        # `until_wall` is not valid, but any time just before is.
+        # So by subtracting a second and adding .999999 seconds
+        # we get the last microsecond just before.
+        before_naive =
+          first_period_until_wall
+          |> Map.put(:microsecond, {999_999, 6})
+          |> NaiveDateTime.add(-1)
+
+        after_naive = second_period_from_wall
+
+        {:ok, latest_datetime_before} =
+          do_from_naive(
+            before_naive,
+            time_zone,
+            first_period.std_offset,
+            first_period.utc_offset,
+            first_period.zone_abbr
+          )
+
+        {:ok, first_datetime_after} =
+          do_from_naive(
+            after_naive,
+            time_zone,
+            second_period.std_offset,
+            second_period.utc_offset,
+            second_period.zone_abbr
+          )
+
+        {:gap, latest_datetime_before, first_datetime_after}
+
+      {:error, _} = error ->
+        error
+    end
+  end
+
+  def from_naive(%{calendar: calendar} = naive_datetime, time_zone, tz_db_or_config)
+      when calendar != Calendar.ISO do
+    # For non-ISO calendars, convert to ISO, create ISO DateTime, and then
+    # convert to original calendar
+    iso_result =
+      with {:ok, in_iso} <- NaiveDateTime.convert(naive_datetime, Calendar.ISO) do
+        from_naive(in_iso, time_zone, tz_db_or_config)
```

Comment:
I think it would be more efficient if in this case we use the TimeZoneClient function that works with iso days, as it would require less conversions. I am playing with it and I will let you know.

Reply:
Actually, the current approach is correct and my suggestion is wrong. We cannot assume the given data is correct, so we do need to convert and validate.